source/byte_buf.c

```diff
@@ -35,8 +35,33 @@ int aws_byte_buf_init(struct aws_byte_buf *buf, struct aws_allocator *allocator,
     return AWS_OP_SUCCESS;
 }
 
-bool aws_byte_buf_is_valid(const struct aws_byte_buf *buf) {
-    return (buf->len <= buf->capacity) && (buf->allocator != NULL) && AWS_MEM_IS_WRITABLE(buf->buffer, buf->len);
+int aws_byte_buf_init_copy(struct aws_byte_buf *dest, struct aws_allocator *allocator, const struct aws_byte_buf *src) {
```

Comment:
Why did we unlink _copy from _copy_from_cursor?  Did that cause proof issues?  This also leads to a slight change in semantics because now capacity mirrors the original buffer's capacity rather than just the length.  

I was about to argue that I preferred the old approach rather than the new approach but then I changed my mind: if you want a shrunk copy use the cursor interface, if you want an exact copy use the byte buf interface.  I would like some additional reviewers to approve that change though.

Reply:
The point to unlink `_copy` from `_copy_from_cursor` is the same that made you change your mind, i.e., if we want a exact copy use `_init_copy`, otherwise just use `_copy_from_cursor`

Reply:
To give some more background: our new proof style involves determining the invariants of the function, and then using CBMC to prove those hold.  For ```init_copy``` we had expected the result to be an exact copy of the initial, and were surprised when its backing array was smaller.  IMO, the new behaviour follows the principle of least surprise.